src/main/java/org/jmxtrans/embedded/EmbeddedJmxTrans.java

```diff
@@ -274,56 +238,98 @@ public String toString() {
 
             shutdownHook = new EmbeddedJmxTransShutdownHook();
             shutdownHook.registerToRuntime();
-            this.state.set(State.STARTED);
+            state = State.STARTED;
             logger.info("EmbeddedJmxTrans started");
         } finally {
-            readWriteLock.writeLock().unlock();
+            lifecycleLock.writeLock().unlock();
         }
     }
 
-
     /**
      * Stop scheduled executors and collect-and-export metrics one last time.
      */
     @PreDestroy
-    public synchronized void stop() throws Exception {
-        readWriteLock.writeLock().lock();
+    public void stop() throws Exception {
+        logger.info("Stop...");
+        lifecycleLock.writeLock().lock();
         try {
-            State state = this.state.get();
             if (!State.STARTED.equals(state)) {
                 logger.debug("Ignore stop() command for " + state + " instance");
                 return;
             }
-            this.state.set(State.STOPPING);
+            logger.info("Unregister shutdown hook");
+            this.shutdownHook.unregisterFromRuntime();
+
+            try {
+                logger.info("Collect metrics...");
+                collectMetrics();
+                logger.info("Export metrics...");
+                exportCollectedMetrics();
+            } catch (RuntimeException e) {
+                logger.warn("Failure to collect and export metrics during stop", e);
+            }
+
+            state = State.STOPPING;
+            logger.info("Set state to {}", state);
+            logger.info("Shutdown collectScheduledExecutor ...");
             try {
                 collectScheduledExecutor.shutdown();
                 try {
                     boolean terminated = collectScheduledExecutor.awaitTermination(getQueryIntervalInSeconds(), TimeUnit.SECONDS);
-                    if (!terminated) {
+                    if (terminated) {
+                        logger.info("collectScheduledExecutor successfully shutdown");
+                    } else {
                         List<Runnable> tasks = collectScheduledExecutor.shutdownNow();
-                        logger.warn("Collect executor could not shutdown in time. Abort tasks " + tasks);
+                        logger.warn("collectScheduledExecutor could not shutdown in time. Abort tasks " + tasks);
                     }
                 } catch (InterruptedException e) {
```

Comment:
Instead of throwing the exception I would opt for Thread.currentThread().interrupt(); http://www.ibm.com/developerworks/library/j-jtp05236/index.html has a nice write up explaining different approaches.